doc/dev/adr/0023-property-support.md

```diff
@@ -0,0 +1,306 @@
+  "contents": [
+    {
+      "@type": "Property",
+      "name": "Foo",
+      "schema": {
+        "@type": [ "Object", "PropertyResult" ],
+        "fields": [
+          {
+            "@type": [ "Field", "PropertyValue" ],
+            "name": "foo",
+            "schema": "integer"
+          },
+          {
+            "@type": [ "Field", "ReadError" ],
+            "name": "propError",
+            "schema": "dtmi:com:example:FooPropertyError;1"
+          }
+        ]
+      }
+    },
+    {
+      "@type": "Property",
+      "name": "Bar",
+      "writable": true,
+      "schema": {
+        "@type": [ "Object", "PropertyResult" ],
+        "fields": [
+          {
+            "@type": [ "Field", "PropertyValue" ],
+            "name": "bar",
+            "schema": "string"
+          },
+          {
+            "@type": [ "Field", "ReadError", "WriteError" ],
+            "name": "propError",
+            "schema": "dtmi:com:example:BarPropertyError;1"
+          }
+        ]
+      }
+    }
+  ]
+}
+```
+
+Note that the "propError" field of Property "Bar" is co-typed with both ReadError and WriteError.
+This is not a requirement.
+Writable Properties may define separate error result types for read and write Commands if desired.
+
+### New adjunct type: Fragmented
+
+The third additive change to the Mqtt extension is support for dynamically itemized Properties via the new adjunct type Fragmented.
+
+The following sample model defines a dynamically itemized, writable Property placeholder named Baz.
+
+```json
+{
+  "@context": [ "dtmi:dtdl:context;4", "dtmi:dtdl:extension:mqtt;4" ],
+  "@id": "dtmi:propertySketch:PropertySketch;1",
+  "@type": [ "Interface", "Mqtt" ],
+  "payloadFormat": "Json/ecma/404",
+  "propertyTopic": "sample/property/{propertyName}/{action}",
+  "contents": [
+    {
+      "@type": [ "Property", "Fragmented" ],
+      "name": "Baz",
+      "writable": true,
+      "schema": {
+        "@type": "Map",
+        "mapKey": {
+          "name": "propKey",
+          "schema": "string"
+        },
+        "mapValue": {
+          "name": "propValue",
+          "schema": "integer"
+        }
+      }
+    }
+  ]
+}
+```
+
+A dynamically itemized Property placeholder is defined using co-type Fragmented and a schema that is a Map whose contents are dynamically added Properties.
+The Map keys are the names of the dynamic Properties, and the Map values are the Properties themselves.
+
+The Fragmented adjunct type indicates that the Map values have analogous constraints to the Property values of normal static Properties.
+Specifically, Map values in a read response are not nullable, and all values in the Map should be returned on each read Command.
+Map values in a write request are optional and nullable:
+
+* If a key is omitted from the Map in a write request, the value is not written.
+* If a key is included and the value is non-null, the dynamic Property is added if not yet present and updated if already present.
```

Comment:
This makes sense for Map - is there any plan to support Array as well, or not at this time since the duplicate story is much trickier or should be fully replaced instead of having this `patch` type operation?

Reply:
No plan for Array.  If the user wants an Array as a Property schema, this will work just like any other Property schema (including a non-Fragmented Map).  But the Fragmented Map is specifically here to support dynamically itemized Properties, which need keys to identify the actual run-time Property names.